include/dali/core/mm/cuda_vm_resource.h

```diff
@@ -0,0 +1,616 @@
+
+#if DALI_USE_CUDA_VM_MAP
+#include "dali/core/mm/memory_resource.h"
+#include "dali/core/mm/detail/free_list.h"
+#include "dali/core/small_vector.h"
+#include "dali/core/bitmask.h"
+#include "dali/core/spinlock.h"
+#include "dali/core/device_guard.h"
+
+namespace dali {
+namespace mm {
+
+namespace test {
+class VMResourceTest;
+}  // namespace test
+
+class cuda_vm_resource : public memory_resource<memory_kind::device> {
+ public:
+  explicit cuda_vm_resource(int device_ordinal = -1,
+                            size_t block_size = 0,
+                            size_t initial_va_size = size_t(4) << 30) {
+    device_ordinal_ = device_ordinal;
+    block_size_ = block_size;
+    initial_va_size_ = initial_va_size;
+
+    configure();
+  }
+  using mem_handle_t = CUmemGenericAllocationHandle;
+
+  ~cuda_vm_resource() {
+    purge();
+  }
+
+  void *try_allocate_from_free(size_t size, size_t alignment) {
+    adjust_params(size, alignment);
+    if (size == 0)
+      return nullptr;
+    lock_guard pool_guard(pool_lock_);
+    // try to get a free region that's already mapped
+    void *ptr = try_get_mapped(size, alignment);
+    if (ptr) {
+      stat_add_allocation(size);
+    }
+    return ptr;
+  }
+
+  void deallocate_no_sync(void *ptr, size_t size, size_t alignment) {
+    deallocate_impl(ptr, size, alignment, false);
+  }
+
+  struct Stat {
+    int allocated_blocks;
+    int peak_allocated_blocks;
+    size_t block_size;
+    size_t allocated_va;
+    size_t curr_allocated;
+    size_t peak_allocated;
+    size_t curr_free;
+    size_t curr_allocations;
+    size_t peak_allocations;
+    size_t total_allocations;
+    size_t total_deallocations;
+    size_t total_unmaps;
+  };
+
+  Stat get_stat() const {
+    lock_guard pool_guard(pool_lock_);
+    return stat_;
+  }
+
+  void clear_stat() {
+    lock_guard pool_guard(pool_lock_);
+    stat_ = {};
+  }
+
+ private:
+  void purge() {
+    lock_guard pool_guard(pool_lock_);
+    mem_lock_guard mem_guard(mem_lock_);
+    for (auto &r : va_regions_)
+      r.purge();
+    va_ranges_.clear();
+  }
+
+  void configure() {
+    if (device_ordinal_ < 0) {
+      CUDA_CALL(cudaGetDevice(&device_ordinal_));
+    }
+    if (block_size_ == 0) {
+      CUdevice device;
+      CUDA_CALL(cuDeviceGet(&device, device_ordinal_));
+      CUDA_CALL(cuDeviceTotalMem(&total_mem_, device));
+      size_t grain = cuvm::GetAddressGranularity();
+      block_size_ = std::max(grain,  // at least grain, for correctness
+                             std::min<size_t>(next_pow2(total_mem_ >> 8),  // capacity-dependent...
+                                              64 << 20));  // ...but capped at 64 MiB
+    }
+    stat_.block_size = block_size_;
+  }
+
+  struct va_region {
+    va_region(cuvm::CUAddressRange range, size_t block_size)
+    : address_range(range), block_size(block_size) {
+      size_t size = address_range.size();
+      size_t blocks_in_range = size / block_size;
```

Comment:
Shouldn't it round up?

Reply:
Or assert `blocks_in_range * block_size == size` ?

Reply:
The latter.